src/java.base/share/classes/jdk/internal/module/ModuleInfoWriter.java

```diff
@@ -46,122 +48,124 @@ public final class ModuleInfoWriter {
-            int flags = opens.modifiers().stream()
-                    .map(OPENS_MODS_TO_FLAGS::get)
-                    .reduce(0, (x, y) -> (x | y));
-            String[] targets = opens.targets().toArray(EMPTY_STRING_ARRAY);
-            mv.visitOpen(opens.source().replace('.', '/'), flags, targets);
-        }
-
-        // uses
-        md.uses().stream().map(sn -> sn.replace('.', '/')).forEach(mv::visitUse);
-
-        // provides
-        for (ModuleDescriptor.Provides p : md.provides()) {
-            mv.visitProvide(p.service().replace('.', '/'),
-                            p.providers()
-                                .stream()
-                                .map(pn -> pn.replace('.', '/'))
-                                .toArray(String[]::new));
+        try {
+            return Classfile.buildModule(
+                ModuleAttribute.of(ModuleDesc.of(md.name()), mb -> {
+                    mb.moduleFlags(modifiersToFlags(md.modifiers()));
+
+                    String vs = md.rawVersion().orElse(null);
+                    if (vs != null) mb.moduleVersion(vs);
+
+                    // requires
+                    for (ModuleDescriptor.Requires r : md.requires()) {
+                        int flags = r.modifiers().stream()
+                                .map(REQUIRES_MODS_TO_FLAGS::get)
+                                .reduce(0, (x, y) -> (x | y));
+                        vs = r.rawCompiledVersion().orElse(null);
+                        mb.requires(ModuleDesc.of(r.name()), flags, vs);
+                    }
+
+                    // exports
+                    for (ModuleDescriptor.Exports e : md.exports()) {
+                        int flags = e.modifiers().stream()
+                                .map(EXPORTS_MODS_TO_FLAGS::get)
+                                .reduce(0, (x, y) -> (x | y));
+                        var targets = e.targets().stream().map(ModuleDesc::of)
+                                .toArray(ModuleDesc[]::new);
+                        mb.exports(PackageDesc.of(e.source()), flags, targets);
+                    }
+
+                    // opens
+                    for (ModuleDescriptor.Opens opens : md.opens()) {
+                        int flags = opens.modifiers().stream()
+                                .map(OPENS_MODS_TO_FLAGS::get)
+                                .reduce(0, (x, y) -> (x | y));
+                        var targets = opens.targets().stream().map(ModuleDesc::of)
+                                .toArray(ModuleDesc[]::new);
+                        mb.opens(PackageDesc.of(opens.source()), flags, targets);
+                    }
+
+                    // uses
+                    md.uses().stream().map(sn -> ClassDesc.of(sn)).forEach(mb::uses);
+
+                    // provides
+                    for (ModuleDescriptor.Provides p : md.provides()) {
+                        mb.provides(ClassDesc.of(p.service()),
+                                        p.providers()
```

Comment:
Formatting nit: align `p.providers()` with the first parameter to `mb.provides` - a little confusing with current indentation.